### PR TITLE
sip: use includes from /System or brewed python instead of Xcode

### PR DESCRIPTION
**Pull request by apjanke**
- [X] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [X] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [X] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [X] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)? *Except for the "Formula name conflicts with caskroom/cask/sip", which is not a regression*

-----

Addresses https://github.com/Homebrew/homebrew-core/pull/18155#issuecomment-330148493. This should prevent `sip` tests from failing when Xcode version changes affect the MacOSX SDK version number/path. And as a salutary effect, it'll make `sip` work under more user configurations.

The default `sip` build bakes in include paths to `/Applications/Xcode.app/.../SDKs/MacOSX10.XXX` in its `lib/python*/site-packages/sipconfig.py` file. This means that it's broken if you install it from a bottle and do not have Xcode installed at the default location, or if you move Xcode after installing it. And it'll break when the MacOSX SDK version increments, as it did with Xcode 9.

This PR changes `sip` to use the paths from `/System/Library/Frameworks` or the brewed Python 2.7 (if installed), removing the `/Applications/Xcode.app/.../SDKs` dependency, for the 2.6/2.7 Pythons.

Note: This PR does not change paths used by Python 3, which means `python3` version bumps will still require a `sip` revision bump. And `sip` builds done against a brewed Python 2 will still have paths to the kegs built in, so `python` revision bumps will still require rebuilds for users who installed `sip` from source against a brewed `python`. I think those concerns are out of scope for this change, though. Or maybe this should go all the way and try to change all `sip`'s path to use `opt_prefix`? I don't know if that would be legit, though, since I don't know if Python preserves the relevant compatibility across minor versions.

### CODE REVIEW

**Comment by ilovezfs** [comment]
LGTM. I'd run `pyqt` and `qscintilla2` through with revision bumps here, even if we don't ship the revision bumps, as a sanity check.

**Comment by apjanke** [comment]
Hmm. Something has changed. In the old `sip` 4.19.3_1 bottle, the paths in `PREFIX/lib/python2.7/site-packages/sipconfig.py` picked up the system `python`.

```
_pkg_config = {
    'arch':               '',
    'default_bin_dir':    '/System/Library/Frameworks/Python.framework/Versions/2.7/bin',
    'default_mod_dir':    '/Library/Python/2.7/site-packages',
    'default_sip_dir':    '/usr/local/share/sip',
    'deployment_target':  '10.12',
    'platform':           'macx-g++',
    'py_conf_inc_dir':    '/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk/System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7',
    'py_inc_dir':         '/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk/System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7',
    'py_lib_dir':         '/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/config',
    'py_version':         0x02070a,
    'qt_framework':       0,
    'sip_bin':            '/usr/local/Cellar/sip/4.19.3_1/bin/sip',
    'sip_config_args':    '--deployment-target=10.12 --destdir=/usr/local/Cellar/sip/4.19.3_1/lib/python2.7/site-packages --bindir=/usr/local/Cellar/sip/4.19.3_1/bin --incdir=/usr/local/Cellar/sip/4.19.3_1/include --sipdir=/usr/local/share/sip',
    'sip_inc_dir':        '/usr/local/Cellar/sip/4.19.3_1/include',
    'sip_mod_dir':        '/usr/local/Cellar/sip/4.19.3_1/lib/python2.7/site-packages',
    'sip_version':        0x041303,
    'sip_version_str':    '4.19.3',
    'universal':          ''
}
```

But in the new 4.19.3_2 bottle from this PR, it's picking up paths to a brewed `python`.

```
_pkg_config = {
    'arch':               '',
    'default_bin_dir':    '/usr/local/Cellar/python/2.7.13_1/Frameworks/Python.framework/Versions/2.7/bin',
    'default_mod_dir':    '/usr/local/Cellar/python/2.7.13_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages',
    'default_sip_dir':    '/usr/local/share/sip',
    'deployment_target':  '10.12',
    'platform':           'macx-g++',
    'py_conf_inc_dir':    '/usr/local/opt/python/Frameworks/Python.framework/Versions/2.7/include/python2.7',
    'py_inc_dir':         '/usr/local/opt/python/Frameworks/Python.framework/Versions/2.7/include/python2.7',
    'py_lib_dir':         '/usr/local/Cellar/python/2.7.13_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/config',
    'py_version':         0x02070d,
    'qt_framework':       0,
    'sip_bin':            '/usr/local/Cellar/sip/4.19.3_2/bin/sip',
    'sip_config_args':    '--deployment-target=10.12 --destdir=/usr/local/Cellar/sip/4.19.3_2/lib/python2.7/site-packages --bindir=/usr/local/Cellar/sip/4.19.3_2/bin --incdir=/usr/local/Cellar/sip/4.19.3_2/include --sipdir=/usr/local/share/sip --configuration=/private/tmp/sip-20170924-19440-1ejm2e0/sip-4.19.3/sip_brew.conf',
    'sip_inc_dir':        '/usr/local/Cellar/sip/4.19.3_2/include',
    'sip_mod_dir':        '/usr/local/Cellar/sip/4.19.3_2/lib/python2.7/site-packages',
    'sip_version':        0x041303,
    'sip_version_str':    '4.19.3',
    'universal':          ''
}
```

(I got that file by doing a `brew pull` on this PR and installing from the bottle, and confirmed it by looking directly at the bottle tarball.)

I believe this will cause breakage with the new `python` revision bump. And it might mean that it effectively requires a brewed `python`, instead of `:python`.

Looking at `brew log sip`, maybe this comes from https://github.com/Homebrew/homebrew-core/pull/15726, which made `:python` `recommended`? But that included a revision bump and a new bottle was pulled, so I would think that that would have ended up with the brewed `python` paths, too.

**Comment by apjanke** [comment]
Okay, I totally redid this with a different approach, using `ENV.delete("SDKROOT")` and keg→opt_prefix path replacement in `sipconfig.py` per @ilovezfs's suggestion. Worked in local testing for me, including with `pyqt` and `qscintilla2`, getting non-Xcode paths picked up in both the default and brewed-python-installed cases. I checked the binary and there are no `HOMEBREW_PREFIX` paths in there; looks like they're contained to `sipconfig.py` and this will be adequate.

What do you think?

Do not merge yet! I've included revision bumps for `pyqt` and `qscintilla2`, just to get them to run through CI for sanity check, per @ilovezfs. I'll remove those in a later amendment before merging this.

File: Formula/sip.rb
Comment (ilovezfs) [inline Formula/sip.rb]: Formula["python"].opt_prefix.realpath

File: Formula/sip.rb
Comment (ilovezfs) [inline Formula/sip.rb]: Formula["python"].opt_prefix

File: Formula/sip.rb
Comment (ilovezfs) [inline Formula/sip.rb]: I think we need the installed check here too since there are other python3s that can satisfy the requirement

File: Formula/sip.rb
Comment (ilovezfs) [inline Formula/sip.rb]: maybe check the location of `python`?

File: Formula/sip.rb
Comment (apjanke) [reply Formula/sip.rb]: Right. Fixed.

**Comment by apjanke** [comment]
Addressed the path abstractions and the is-python3-installed check. Lemme think about checking the location of `python`.

File: Formula/sip.rb
Comment (DomT4) [reply Formula/sip.rb]: What are we doing here if the user isn't using Homebrew's `python`? The Requirement searches the whole `$PATH` right, so if building from source they could plausibly want something that is neither brew nor system `python`, no?

Comment (ilovezfs) [reply Formula/sip.rb]: I believe the requirement currently strictly prefers Homebrew's Python if it's installed.

Comment (DomT4) [reply Formula/sip.rb]: But presumably if it isn't this could cause some unexpected behaviour? I mean I guess you/Mike may argue that Homebrew is essentially not a from-source package manager any more, so maybe the discussion point is effectively nullified by that.

Comment (ilovezfs) [reply Formula/sip.rb]: I don't understand the point. If `Formula["python"].installed?` is true and the require always uses the formula if it's installed, then what are you asking?

Comment (DomT4) [reply Formula/sip.rb]: Oh, crap. I can't read. Sorry, I thought an earlier version of this PR was hardcoding in `/System` in the event Homebrew's `python` wasn't found & so the choice was essentially reduced to either system or brew & nothing else, even when building from source, and I guess I got hung up on that :see_no_evil:.

Forgive me, I slept for 3 hours before someone two floors above filled their kitchen with smoke & forced a 45 minute evacuation this morning.

Comment (ilovezfs) [reply Formula/sip.rb]: Sounds fun.

Of course, this could just use `inreplace "file", "foo", "bar", false` and call it a day.

Comment (DomT4) [reply Formula/sip.rb]: Or fixing the `python-config` scripts & retaining just the `ENV.delete("SDKROOT")` here, presumably? I don't know. Maybe I should stop making suggestions for the day :laughing:.

Comment (ilovezfs) [reply Formula/sip.rb]: >Or "fixing" the python-config scripts

Yes, I suggested that to apjanke yesterday.

Comment (DomT4) [reply Formula/sip.rb]: We do the same thing in a bunch of places; I think it'd be a reasonable step, personally.

**Comment by ilovezfs** [comment]
@apjanke I think we can just do the `ENV.delete("SDKROOT")` here and then worry about the Cellar -> opt thing by fixing `python2_config` and `python3_config` at some point, rather than making changes in `sip` to hack around the issues with the `python` and `python3` formulae.

It will still require `sip` to get revision bumped in the python PRs until that's fixed, but at least no more spontaneous path changes when Xcode gets upgraded.

**Comment by apjanke** [comment]
Sounds good to me. I have amended this PR to do just the `ENV.delete("SDKROOT")`.

**Review by ilovezfs — APPROVED** [review]

**Comment by ilovezfs** [comment]
:rocket:

**Pull request closed without merging**
